### PR TITLE
Extract teardowns to ensure they run on failed tests

### PR DESCRIPTION
**Pull request by DanTup**
Otherwise we leak flutter_tester processes that can cause the rest of the tests to fail.

(Similar to #214).

### CODE REVIEW

**Comment by DanTup** [comment]
This diff looks a little weird unless you ignore whitespace and view side-by-side so it's easier to review here:

https://github.com/flutter/devtools/pull/215/files?utf8=%E2%9C%93&diff=split&w=1

I had to add empty groups to handle that some tests have different setup code. Possibly the teardown could come up to the top level, but it felt better to keep them symmetrical.

**Comment by DanTup** [comment]
Actually, these empty groups don't get collapsed like I thought they would:

tree

@jacob314 I think we could assign reasonable names to these groups - one is starting with defaults (eg `trackWidgetCreation: true`) and the other stats with trackWidetCreation off (even though only one of them makes it clear from the name). Happy for me to add names and leave the groups in?

**Comment by DanTup** [comment]
Hmm, something went bad... taking a look

**Comment by DanTup** [comment]
I've reverted part of these changes do avoid the empty groups and trying to share setups (it'd become a pain if we needed further customisations), so now this just shares teardowns (as well as bailing from teardown if it doesn't need to do anything).

File: packages/devtools/test/support/flutter_test_environment.dart [hunk 1]
```diff
@@ -42,6 +42,7 @@ class FlutterTestEnvironment {
   set beforeTearDown(VoidAsyncFunction f) => _beforeTearDown = f;
 
   bool _needsSetup = true;
+  bool get _needsTeardown => !_needsSetup;
```
Comment (jacob314) [inline packages/devtools/test/support/flutter_test_environment.dart]: nit: remove this flag. I think checking _needsSetup directly is fine and a little clearer

**Comment by jacob314** [comment]
The teardown case is the part we need for robustness so only worrying about it is fine. lgtm

File: packages/devtools/test/support/flutter_test_environment.dart (hunk 1 above)
Comment (DanTup) [reply packages/devtools/test/support/flutter_test_environment.dart]: This was an overlap with https://github.com/flutter/devtools/pull/214. I'll rebase on top of that once it's merged and ensure this is gone. Thanks!

**Pull request merged**